### PR TITLE
fix: allow for more flexibility for custom components? and using the enhanced object dialog

### PR DESCRIPTION
**Pull request by RitaDias**
### Description

- This PR allows for more flexible handling of custom components with using the enhanced object dialog with minimal outside impact
- Update to comments have been added to clarify usage
- New updated tests for custom components that have caused the enhacned object dialog to malfunction in the past
- New automated tests for potential pitfalls: keyboard navigation in items and creating new elements and closing modal

Before

https://github.com/user-attachments/assets/82798bc5-2f1f-4dec-b4a7-afb532b4a26f

After

https://github.com/user-attachments/assets/b1c3b8ff-9e66-4012-a142-c82a6762d182

### What to review

This approach has been discussed with other folks in the team, the solution here checks if an input has been added which matches the current focus path - these input determinators are provided automatically by us and show allow for flexibility when using renderDefaults

### Testing

All tests should pass.
New automated tests have been added 
Manual tests have been done including with documents that are side by side as to attempt to narrow down edge cases

### Notes for release

Enhanced object dialog (beta) improvements: added more flexible approach for custom components within array of objects. Custom components should open as expected.

### CODE REVIEW

File: packages/sanity/src/core/form/useDocumentForm.ts [hunk 1]
```diff
@@ -528,12 +529,26 @@ export function useDocumentForm(options: DocumentFormOptions): DocumentFormValue
         if (!isKeySegment(lastSegment)) {
           // For fields inside array items, find the last key segment to preserve context
           const lastKeyIndex = nextFocusPath.findLastIndex((seg) => isKeySegment(seg))
+          // We must preserve the context of the array item when focusing on an object field
+          // As if we don't, then the open path will open the dialog when the object is simply focused,
+          // Which is not what we want.
```
Comment (RitaDias) [inline packages/sanity/src/core/form/useDocumentForm.ts:534]: Clarifying what the logic did and what it was preventing

Comment (RitaDias) [reply packages/sanity/src/core/form/useDocumentForm.ts:534]: Updated tests since initially I thought I might need to. change this logic and it made a silent regression happen (clicking a new object item, writing in the field and closing it was requiring two clicks and, a different bug where tabbing through object of arrays was causing the dialog to open automically, this was, however, due to my previous change and no longer happens). In any case, there are tests for this now

File: e2e/tests/enhanced-object-dialog/smoke.spec.ts
Comment (RitaDias) [inline e2e/tests/enhanced-object-dialog/smoke.spec.ts:1]: Updated the close smoke test to actually create the element before closing as this makes sure that the health of the logic in the useDocumentForm is kept up.

Comment (RitaDias) [reply e2e/tests/enhanced-object-dialog/smoke.spec.ts:1]: Added new tests to make sure keyboard navigation in array of objects works as expected

File: packages/sanity/src/core/form/useDocumentForm.ts [hunk 2]
```diff
@@ -528,12 +529,26 @@ export function useDocumentForm(options: DocumentFormOptions): DocumentFormValue
...
+           * This checks if an element exists in the dom with the updated path.
+           * There are situations where the element doesn't exist in the dom yet, for example:
+           * (think internationalizedArrayString where clicking a button creates an input with the path "en.value")
+           * And, in those cases, the open path will open the dialog when the button is pressed.
+           * So, if an element exists in the dom with the updated path then it means
+           * then we shouldn't update the open path
+           */
+          const elementExists = document.getElementById(`${pathToString(nextFocusPath)}`)
+          if (!elementExists) {
+            handleSetOpenPath(pathFor(newOpenPath))
```
Comment (RitaDias) [inline packages/sanity/src/core/form/useDocumentForm.ts:550]: This logic existed within the enhancedObjectDialog, however, even in an instance where we passed the focusPath in a prop, the path and props we received there were not updated soon enough and so it means it doesn't have the flexibility that some of our use cases required. I have tried other approaches and this place has been the only one I've found so far that can tackle it to the level we require.

Comment (RitaDias) [reply packages/sanity/src/core/form/useDocumentForm.ts:550]: fyi @juice49 @bjoerge
